src/klein/_request.py

```diff
@@ -0,0 +1,341 @@
+        """
+
+
+    def bodyAsBytes():
+        # type: () -> Deferred[bytes]
+        """
+        The request entity body, as bytes.
+
+        @note: This necessarily reads the entire request body into memory,
+            which may be a problem if the body is large (eg. a large upload).
+
+        @note: This method caches the body, which means that unlike
+            C{self.bodyAsFount}, calling it repeatedly will return the same
+            data.
+
+        @note: This method accesses the fount (via C{self.bodyAsFount}), which
+            means the found will not be available afterwards, and that if
+            C{self.bodyAsFount} has previously been called directly, this
+            method will raise L{NoContentError}.
+
+        @raise NoContentError: If the fount has previously been accessed.
+        """
+
+
+
+# Simple implementation
+
+def validateBody(instance, attribute, body):
+    # type: (Any, Any, Union[bytes, IFount]) -> None
+    if (
+        type(body) is not bytes and
+        not IFount.providedBy(body)
+    ):
+        raise TypeError("body must be bytes or IFount")
+
+
+@implementer(IHTTPRequest)
+@attrs(frozen=True)
+class FrozenHTTPRequest(object):
+    """
+    Immutable HTTP request.
+    """
+
+    @attrs(frozen=False)
+    class _State(object):
+        """
+        Internal mutable state for L{HTTPRequestFromIRequest}.
+        """
+
+        _cachedBody = attrib(
+            validator=optional(instance_of(bytes)), default=None, init=False
+        )  # type: bytes
+
+    method  = attrib(validator=instance_of(Text))  # type: Text
+    uri     = attrib(validator=instance_of(URL))  # type: URL
+    headers = attrib(
+        validator=instance_of(FrozenHTTPHeaders)
+    )  # type: FrozenHTTPHeaders
+
+    _body = attrib(validator=validateBody)  # type: Union[bytes, IFount]
+
+    _state = attrib(default=Factory(_State), init=False)  # type: _State
+
+
+    def bodyAsFount(self):
+        # type: () -> IFount
+        if type(self._body) is bytes:
+            return bytesToFount(self._body)
+
+        # assuming: IFount.providedBy(self._body)
+
+        return self._body
+
+
+    def bodyAsBytes(self):
+        # type: () -> bytes
+        if type(self._body) is bytes:
+            return succeed(self._body)
+
+        # assuming: IFount.providedBy(self._body)
+
+        def cache(bodyBytes):
+            # type: (bytes) -> bytes
+            self._state._cachedBody = bodyBytes
+            return bodyBytes
+
+        d = fountToBytes(self._body)
+        d.addCallback(cache)
+        return d
+
+
+
+# Support for L{IRequest}
+
+@implementer(IHTTPRequest)
+@attrs(frozen=True)
+class HTTPRequestFromIRequest(object):
+    """
+    HTTP request.
+
+    This is an L{IHTTPRequest} implementation that wraps an L{IRequest} object.
+
+    This is used by Klein to expose objects from L{twisted.web} to clients
+    while presenting the Klein interface.
+    """
+
+    @attrs(frozen=False)
+    class _State(object):
```

Comment:
This looks like it's practically screaming for https://github.com/python-attrs/attrs/issues/133 :)